### ISSUE TITLE
Aggregated API servers are unexpectedly throttled when making anonymous requests due to the `UnauthenticatedHTTP2DOSMitigation` feature gate

### ISSUE THREAD

**Issue by benjaminapetersen** [kubernetes/kubernetes#122308]
### What happened?

When enabling the UnauthenticatedHTTP2DOSMitigation feature flag on our aggregated API server in the [Pinniped project](https://github.com/vmware-tanzu/pinniped), we began to experience a number of flakes in our CI due to the throttling of anonymous requests introduced by this feature gate.  This is likely an unintended consequence, the feature gate should not affect aggregated API servers. 

/kind bug
/kind regression
/assign enj
/sig auth
/sig api-machinery

### What did you expect to happen?

We expected public anonymous requests to be throttled, not anonymous requests in the conversation between our API server and Kube API server. 

### How can we reproduce it (as minimally and precisely as possible)?

We attempted to prove the problem by modifying an integration test.  See the diff here:
https://github.com/kubernetes/kubernetes/compare/master...benjaminapetersen:ben/aggregated-api-server/feature-gate-problem?expand=1

However, this change does not successfully prove the issue.  We believe it may be because the test fakes out too many components to accurately reproduce the error. 

### Kubernetes version

<details>

```console
$ kubectl version
# n/a ~ upcoming release 
```

</details>

### Cloud provider

<details>
n/a
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

cc @cfryanr for significant effort investigating this issue.

**Comment by cfryanr** [comment]
Here is some more information about how I was able to reliably reproduce the problem using the Pinniped Concierge app, which is an aggregated API server. https://github.com/vmware-tanzu/pinniped/pull/1813#issuecomment-1854427318

I suspect that any similar aggregated API implementation which allows unauthenticated users to make requests to the API would suffer the same problem, when the UnauthenticatedHTTP2DOSMitigation feature gate is turned on in the aggregated API server’s process. Note that the UnauthenticatedHTTP2DOSMitigation feature does not need to be turned on for the Kubernetes API server itself to reproduce the problem. 

My concern is that if a future version of the Kubernetes libraries enables this feature gate by default and also removes any way to disable it, then aggregated API servers implemented in golang using the Kubernetes libraries won’t reliably be able to serve APIs for anonymous users anymore. 

It seems like maybe the Kube API server should more carefully respect http2 goaway on responses coming from an aggregated API server?

**Comment by sftim** [comment]
(nit)
/retitle Aggregated API servers are unexpectedly throttled when making anonymous requests due to the `UnauthenticatedHTTP2DOSMitigation` feature gate

**Comment by sftim** [comment]
Maybe the aggregated API server could make an authenticated request and use impersonation to turn the request anonymous?

**Comment by enj** [comment]
@cfryanr @benjaminapetersen can you cherrypick #122322 into pinniped and see if it fixes the issue when `UnauthenticatedHTTP2DOSMitigation` is enabled?

**Comment by enj** [comment]
> It seems like maybe the Kube API server should more carefully respect http2 goaway on responses coming from an aggregated API server?

@cfryanr This isn't possible AFAIK because the connection close can happen in the middle of any request, and the Kube API server has no way to know if retrying the request is safe to do.

> Maybe the aggregated API server could make an authenticated request and use impersonation to turn the request anonymous?

@sftim this problem is anonymous user -> KAS -> aggregated API server.

**Comment by sftim** [comment]
Ah, the _aggregated API server_ is mitigating the DoS?  Does it need to?

**Comment by enj** [comment]
> Ah, the _aggregated API server_ is mitigating the DoS? Does it need to?

Yes, in the same way that it is required to re-authenticate and re-authorize the request.  It is running as a Kube service on the cluster, and cannot make any assumptions about how the request got to it.  The code paths for authentication are shared between KAS and aggregated API servers (as they should be), and I fully expect to drop the feature gate and always have the DOS mitigation enabled in some future release (meaning KAS and all aggregated API servers using the new library code would enforce the mitigation).

**Comment by sftim** [comment]
I guess the kube-apiserver could add a header to indicate the provenance. Like `Via:` but if you could trust the value.

> The Kubernetes apiserver connects to the extension apiserver over TLS, authenticating itself using a client certificate.

_&mdash; Kubernetes docs_

The aggregated API server _can_ assume that a request made using that client certificate has arrived via the kube-apiserver and may wish to assume that the kube-apiserver has already done DoS mitigation.

---

PS. W don't actually know the aggregated API server is running in-cluster.

**Comment by cfryanr** [comment]
> @cfryanr @benjaminapetersen can you cherrypick https://github.com/kubernetes/kubernetes/pull/122322 into pinniped and see if it fixes the issue when UnauthenticatedHTTP2DOSMitigation is enabled?

Hi @enj,

I performed the following steps:
1. Create a branch in my locally cloned k/k repo from the `v1.28.4` tag and checkout that branch.
2. Compile our aggregated API server using a `replace` directive to replace `k8s.io/apiserver` with my local k/k repo's `staging/src/k8s.io/apiserver` directory.
3. Deploy to a Kind cluster running Kube v1.27.3.
4. Confirm that I can still reproduce the problem (without your PR commit) by making lots of parallel anonymous requests to the aggregated API by using curl against the Kube API server. I could see that some requests were failing with 503 responses from the Kube API server, with response bodies that say `error trying to reach service: http2: Transport: cannot retry err [http2: Transport received Server's graceful shutdown GOAWAY] after Request.Body was written; define Request.GetBody to avoid this error`.
5. Cherry pick your commit from your PR onto my local k/k branch.
6. Recompile and redeploy onto the Kind cluster.
7. Attempt to reproduce the problem again the same way.

With your commit added on top of the `k8s.io/apiserver` v1.28.4 package, I was not able to reproduce the problem anymore. I could successfully make many simultaneous anonymous requests to the aggregated API, and I did not see the error at all.

**Comment by enj** [comment]
> The aggregated API server _can_ assume that a request made using that client certificate has arrived via the kube-apiserver and may wish to assume that the kube-apiserver has already done DoS mitigation.

@sftim this is exactly the approach that I took in #122322.

> With your commit added on top of the `k8s.io/apiserver` v1.28.4 package, I was not able to reproduce the problem anymore. I could successfully make many simultaneous anonymous requests to the aggregated API, and I did not see the error at all.

@cfryanr thanks for confirming.  That tells me that my understanding of the root cause of the issue is correct.  I will work with the other maintainers early next year to see if they agree with the approach (and to add a test that can reproduce the issue so we do not regress again).

**Comment by cfryanr** [comment]
Thanks @enj!!

**Comment by cfryanr** [comment]
Just checking in after a few months. Any update on this @enj?

Ideally this issue would block the UnauthenticatedHTTP2DOSMitigation feature gate from progressing to becoming non-configurable, until this is fixed.

**Comment by enj** [comment]
@cfryanr it is too late in the v1.30 cycle for #122322 to land, so v1.31 is the earliest I see us fixing this.  I don't see us promoting `UnauthenticatedHTTP2DOSMitigation` until then.

**Comment by cfryanr** [comment]
Thanks for the update @enj!

**Comment by cfryanr** [comment]
@enj The bot says that this issue needs to be re-triaged. Still planning on eventually fixing it? 🤞

**Comment by cfryanr** [comment]
/reopen

I haven't heard that this issue was fixed. Can it be reopened?

**Comment by cfryanr** [comment]
Thanks @enj!
